### PR TITLE
Torchvision scaler

### PR DESCRIPTION
**Pull request by corey-nm**
Properly passes scaler to wrap_optim.

# Test plan

Ran
- without `--amp`
- with `--amp`

**Have not tested amp with quantization**

### CODE REVIEW

File: src/sparseml/pytorch/torchvision/train.py [hunk 1]
```diff
@@ -561,9 +561,17 @@ def log_metrics(tag: str, metrics: utils.MetricLogger, epoch: int, epoch_step: i
...
+            model,
+            optimizer,
+            steps_per_epoch=steps_per_epoch,
+            epoch=args.start_epoch,
+            wrap_optim=scaler,
         )
+        if scaler is None:
+            optimizer = step_wrapper
+        else:
+            scaler = step_wrapper
```
Comment (corey-nm) [inline src/sparseml/pytorch/torchvision/train.py:574]: Will re-ask this question here:

If using quantization & amp, the scaler will be set to None down below. This means that train_one_epoch will directly using optimizer instead of scaler. However the manager.modify wrapped the scaler.

Does that mean I need to re-call manager.modify on the optimizer after amp is disabled?

Comment (corey-nm) [reply src/sparseml/pytorch/torchvision/train.py:574]: Or should I disable scaler by doing `scaler._enabled = False`? That seems to be what transformers integration does

Comment (corey-nm) [reply src/sparseml/pytorch/torchvision/train.py:574]: Resolved in meeting - setting enabled flag for scaler down below

Comment (KSGulin) [reply src/sparseml/pytorch/torchvision/train.py:574]: As mentioned offline, just setting `scaler._enabled = False` is enough. The optimizer step will be taken care of by torch here:
https://github.com/pytorch/pytorch/blob/d8e795ecd53670682bd3b2e5ff1f378402b147d5/torch/cuda/amp/grad_scaler.py#L313

**Review by KSGulin — APPROVED** [review]

**Review by bfineran — APPROVED** [review]
nice

File: src/sparseml/pytorch/torchvision/train.py [hunk 2]
```diff
@@ -568,9 +568,17 @@ def log_metrics(tag: str, metrics: utils.MetricLogger, epoch: int, epoch_step: i
             loggers=logger,
             distillation_teacher=args.distill_teacher,
         )
-        optimizer = manager.modify(
-            model, optimizer, steps_per_epoch=steps_per_epoch, epoch=args.start_epoch
+        step_wrapper = manager.modify(
```
Comment (natuan) [inline src/sparseml/pytorch/torchvision/train.py:571]: @corey-nm We should pass the loggers into here so that from now on logging could be handled both by the modifiers (relevant for e.g. logging different quantization losses) and by the loss update of the wrapped manager (i.e. the RecipeManagerStepWrapper). It also prepares us well for distillation in the future.

I think we also need to recompute the steps per epoch passed in, taking into account grad accum steps and multi gpus.  Look at this line and those right above: https://github.com/neuralmagic/sparseml/blob/main/src/sparseml/transformers/sparsification/trainer.py#L272

After that, we can remove all the logging from this script (except perhaps for the python logger).

**Review by natuan — CHANGES_REQUESTED** [review]

File: src/sparseml/pytorch/torchvision/train.py (hunk 2 above)
Comment (natuan) [reply src/sparseml/pytorch/torchvision/train.py:571]: Btw, looks like we can unify scaler and optimizer into "step_wrapper" and just pass it into the `train_one_epoch`(instead of having 2 separate/mutually exclusive objects)?

Comment (corey-nm) [reply src/sparseml/pytorch/torchvision/train.py:571]: 1. the loggers were passed into manager.initializer() in the call right above this - does that not pass the loggers to all the modifiers?
2. steps_per_epoch does take into account gradient accumulation, but not multi gpu. i can fix the multi gpu in a separate pr
3. yeah the scaler/optimizer interactions are weird. you need to pass optimizer to scaler.step, and you also need to call optimizer.zero_grad() through the optimizer. it seems like if they were designed differently they could be combined. we could create a new class that wraps them, but i'm not sure the indirection it would add would be worth it. thoughts?

Comment (natuan) [reply src/sparseml/pytorch/torchvision/train.py:571]: I see, you're right about the loggers init. 
Thanks for taking on 2. For 3, no I didn't mean to push for it unless it's an easy clean up.

Comment (natuan) [reply src/sparseml/pytorch/torchvision/train.py:577]: I think let's push this in and continue to improve it

**Review by natuan — APPROVED** [review]
